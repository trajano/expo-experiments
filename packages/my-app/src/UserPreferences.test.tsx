import { act, fireEvent, render } from '@testing-library/react-native';
import { useCallback } from 'react';
import { Button, Text } from 'react-native';
import {
  UserPreferencesProvider,
  useUserPreferences,
  WithUserPreferences,
} from './UserPreferences';
import AsyncStorage from '@react-native-async-storage/async-storage';
jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);
const TestComponent: React.FC = () => {
  const { preferences, setAsync } = useUserPreferences();
  const setSomething = useCallback(async () => {
    await setAsync('foo', 'bar');
  }, [setAsync]);
  const setSomethingElse = useCallback(async () => {
    await setAsync('foo', 'baz');
  }, [setAsync]);
  const setSomethingNull = useCallback(async () => {
    await setAsync('foo', null);
  }, [setAsync]);
  return (
    <>
      <Text testID="props">{JSON.stringify(preferences)}</Text>
      <Button
        onPress={setSomething}
        title="setSomething"
        testID="setSomething"
      />
      <Button
        onPress={setSomethingElse}
        title="setSomethingElse"
        testID="setSomethingElse"
      />
      <Button
        onPress={setSomethingNull}
        title="setSomethingNull"
        testID="setSomethingNull"
      />
    </>
  );
};

describe('UserPreferences', () => {
  beforeEach(async () => {
    await AsyncStorage.clear();
  });
  it('Use provider and update', async () => {
    const { getByTestId, toJSON } = render(
      <UserPreferencesProvider
        userPreferencesInitial={{}}
        userPreferencesStorageKey="storage"
      >
        <TestComponent />
      </UserPreferencesProvider>,
    );
    expect(await AsyncStorage.getItem('storage')).toBeNull();
    await act(() => Promise.resolve());
    expect(await AsyncStorage.getItem('storage')).toEqual('{}');
    expect(getByTestId('props').props.children).toBe(JSON.stringify({}));
    await act(() => fireEvent.press(getByTestId('setSomething')));
    expect(getByTestId('props').props.children).toBe(
      JSON.stringify({ foo: 'bar' }),
    );
    expect(await AsyncStorage.getItem('storage')).toBe(
      JSON.stringify({ foo: 'bar' }),
    );
    await act(() => fireEvent.press(getByTestId('setSomethingElse')));
    expect(await AsyncStorage.getItem('storage')).toBe(
      JSON.stringify({ foo: 'baz' }),
    );
    await act(() => fireEvent.press(getByTestId('setSomethingNull')));
    expect(await AsyncStorage.getItem('storage')).toBe(JSON.stringify({}));
    expect(toJSON()).toMatchSnapshot();
  }, 10_000);

  it('Use provider with initial state and update', async () => {
    const initialState = { foo: 'abc' };
    const { getByTestId, toJSON } = render(
      <UserPreferencesProvider
        userPreferencesInitial={initialState}
        userPreferencesStorageKey="storage"
      >
        <TestComponent />
      </UserPreferencesProvider>,
    );
    expect(await AsyncStorage.getItem('storage')).toBeNull();
    await act(() => Promise.resolve());
    expect(await AsyncStorage.getItem('storage')).toEqual(
      JSON.stringify(initialState),
    );
    expect(getByTestId('props').props.children).toBe(
      JSON.stringify(initialState),
    );
    await act(() => fireEvent.press(getByTestId('setSomething')));
    expect(getByTestId('props').props.children).toBe(
      JSON.stringify({ foo: 'bar' }),
    );
    expect(await AsyncStorage.getItem('storage')).toBe(
      JSON.stringify({ foo: 'bar' }),
    );
    await act(() => fireEvent.press(getByTestId('setSomethingElse')));
    expect(await AsyncStorage.getItem('storage')).toBe(
      JSON.stringify({ foo: 'baz' }),
    );
    await act(() => fireEvent.press(getByTestId('setSomethingNull')));
    expect(await AsyncStorage.getItem('storage')).toBe(
      JSON.stringify(initialState),
    );
    expect(toJSON()).toMatchSnapshot();
  }, 10_000);

  it('Use HoC', async () => {
    const TestedComponent = WithUserPreferences(TestComponent);
    const { getByTestId, toJSON } = render(
      <TestedComponent
        userPreferencesInitial={{}}
        userPreferencesStorageKey="storage"
      />,
    );
    await act(() => Promise.resolve());
    expect(getByTestId('props').props.children).toBe(JSON.stringify({}));
    expect(toJSON()).toMatchSnapshot();
  });
});
