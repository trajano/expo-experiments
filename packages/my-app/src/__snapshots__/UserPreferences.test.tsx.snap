// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPreferences Use HoC 1`] = `
[
  <Text
    testID="props"
  >
    {}
  </Text>,
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="setSomething"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        setSomething
      </Text>
    </View>
  </View>,
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="setSomethingElse"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        setSomethingElse
      </Text>
    </View>
  </View>,
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="setSomethingNull"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        setSomethingNull
      </Text>
    </View>
  </View>,
]
`;

exports[`UserPreferences Use provider 1`] = `
[
  <Text
    testID="props"
  >
    {}
  </Text>,
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="setSomething"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        setSomething
      </Text>
    </View>
  </View>,
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="setSomethingElse"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        setSomethingElse
      </Text>
    </View>
  </View>,
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="setSomethingNull"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        setSomethingNull
      </Text>
    </View>
  </View>,
]
`;
